### PR TITLE
Account Editable Component

### PR DESCRIPTION
**Pull request by qbob1**
This branch represent the Editable fields of the Account page
The process of editing is two fold:
1. Editable Component
- Takes in a data bound html element. When edit mode is toggled and a proper input is provided
it sends an update event emitter.
- Emission is caught by Parent component and sent to the editing service
2. Editing Service
- Light weight state management library that takes in T and manages updates to T
- Current use case uses a partial as T in order to allow for updates to sub-properties of the model object
- Editing service has a property payloadEmitter which is an observable of the state updates.
- Current use is to catch the payload and send put requests to server

Recent Updates:
Editable service is now injected via a const InjectionToken
Unsure of Injection Token file position, please advise

### CODE REVIEW

File: angular/src/app/modules/account/profile/profile.component.html [hunk 1]
```diff
@@ -6,11 +6,17 @@
...
             <img alt="placeholder image" class="is-rounded" src="https://bulma.io/images/placeholders/256x256.png">
           </figure>
-
           <br>
           <div>
             <p class="heading">adult</p>
-            <p class="title is-6">{{ profile.name.given }} {{ profile.name.family }}</p>
+            <p class="title is-6">
+              <uic-editable [(data)]='profile.name.given' Type="name">
+                <p class="title is-6"> {{profile.name.given}} </p>
```
Comment (escalonn) [inline angular/src/app/modules/account/profile/profile.component.html:14]: double quotes & spacing inside `{{ }}`

File: angular/src/app/modules/account/payment/payment.component.ts [hunk 2]
```diff
@@ -1,14 +1,34 @@
...
-  @Input() payment: Payment;
+  @Input() payments: Payment[];
+  @Output() paymentsEdited = new EventEmitter();
 
-  constructor() {}
+  /**
+   * Represents the _Payment Component_ 'constructor' method
+   * @param editingservice AccountEditingService
+   */
+  constructor(private readonly editingservice: AccountEditingService) {}
```
Comment (escalonn) [inline angular/src/app/modules/account/payment/payment.component.ts:20]: camelCase on editingservice, here & elsewhere

File: angular/src/app/modules/account/profile/profile.component.spec.ts [hunk 3]
```diff
@@ -22,6 +23,7 @@ describe('ProfileComponent', () => {
   beforeEach(async(() => {
     TestBed.configureTestingModule({
       declarations: [ProfileComponent],
+      providers: [AccountEditingService],
```
Comment (escalonn) [inline angular/src/app/modules/account/profile/profile.component.spec.ts:26]: unit test

File: angular/src/app/modules/account/services/abstract.editing.service.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+  public editUpdates = this.subject.pipe(
+    scan((acc, curr) => (typeof curr === 'object' ? Object.assign({}, acc, curr) : null), {})
+  );
+
+  public PayloadEmitter = this.editUpdates
+    .pipe(filter((el) => this.typeGuard(el || {})))
+    .pipe(skip(1));
+
+  public update(e: any) {
```
Comment (escalonn) [inline angular/src/app/modules/account/services/abstract.editing.service.ts:15]: all of this `any` should be replaceable by generics...

File: angular/src/app/modules/account/services/account.editing.service.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import { Injectable } from '@angular/core';
+import { EditingService } from './abstract.editing.service';
+import { TypeGuard } from './account.typeguard';
+@Injectable()
+export class AccountEditingService extends EditingService<Account> {
+  constructor() {
```
Comment (escalonn) [inline angular/src/app/modules/account/services/account.editing.service.ts:6]: dependency injection of the typeguard by the module should even remove the need for this separate class. you should define a type (probably either interface or class) for the typeguard

File: angular/src/app/services/account/account.service.spec.ts [hunk 6]
```diff
@@ -25,7 +25,7 @@ describe('AccountService', () => {
     payments: [
       {
         id: 'string',
-        cardExpirationDate: new Date(),
+        cardExpirationDate: 'new Date()',
```
Comment (escalonn) [inline angular/src/app/services/account/account.service.spec.ts:28]: confusing

File: angular/src/app/modules/account/services/account.typeguard.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
+import { Account } from '../../../data/account.model';
+
+export const TypeGuard = (el: any): el is Account => {
```
Comment (escalonn) [inline angular/src/app/modules/account/services/account.typeguard.ts:3]: there really wasnt a good library to handle most of this?
if not, then this sort of thing is fine, although it belongs in a separate PR, since it has to do with data coming from the backend, not data being edited by the user; and because it's much of this is not specific to account.

but camelCase for all these variables and functions, TitleCase only for types/classes/interfaces

File: angular/src/app/modules/account/services/account.typeguard.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    CheckAddress(el.address) &&
+    AllAreTrue(el.profiles, CheckProfile) &&
+    AllAreTrue(el.payments, CheckPayment)
+  ) {
+    return true;
+  } else {
+    return false;
+  }
+};
+export function AllAreTrue(x: Array<any>, CheckFunction: (p: any) => boolean) {
```
Comment (escalonn) [inline angular/src/app/modules/account/services/account.typeguard.ts:18]: what's the need for this, don't arrays have `every` method?

File: angular/src/app.config.json [hunk 9]
```diff
@@ -19,9 +19,18 @@
...
+        "url": "account"
+      },
+      {
+        "text": "booking",
+        "url": "booking"
+      },
+      {
+        "text": "lodging",
+        "url": "lodging"
+      }
```
[the remark this answers is not in the document]
Comment (pamelasoulis) [reply angular/src/app.config.json]: This comment was addressed in #42

File: angular/src/app/modules/account/account-routing.module.ts [hunk 10]
```diff
@@ -7,7 +7,9 @@ import { AddressComponent } from './address/address.component';
 import { PaymentComponent } from './payment/payment.component';
 import { ProfileComponent } from './profile/profile.component';
 import { AccountReviewComponent } from './account-review/account-review.component';
-
+import { EditableComponent } from './editable/editable.component';
+import { FormsModule, ReactiveFormsModule } from '@angular/forms';
+import { AccountEditingService } from './services/account.editing.service';
 const routes: Routes = [{ component: AccountComponent, path: '' }];
```
[the remark this answers is not in the document]
Comment (pamelasoulis) [reply angular/src/app/modules/account/account-routing.module.ts:14]: Addressed in #42

File: angular/src/app/modules/account/account/account.component.html [hunk 11]
```diff
@@ -1,20 +1,22 @@
...
-            <div class="tile box has-background-dark is-child">
+            <div class="tile box has-background-dark is-child" id="addressCard" style="position: relative; z-index: 1;">
               <h2 class="title has-text-light is-4">address</h2>
-              <uic-address *ngIf="address$ | async as address" [address]="address"></uic-address>
+              <uic-address *ngIf="address$ | async as address" [address]="address">
+              </uic-address>
             </div>
 
-            <div class="tile box has-background-dark is-child">
+            <div class="tile box has-background-dark is-child" style="width: 280px;">
```
[the remark this answers is not in the document]
Comment (pamelasoulis) [reply angular/src/app/modules/account/account/account.component.html]: Addressed the declaration in #42

File: angular/src/app/modules/account/address/address.component.html [hunk 12]
```diff
@@ -3,13 +3,16 @@
     <div class="level-item has-text-light">
       <div>
         <p class="heading">street</p>
-        <p class="title has-text-light is-6">{{ address?.street }}</p>
+        <uic-editable [(data)]='address.street' (dataChange)="edited()" Type="street">
```
[the remark this answers is not in the document]
Comment (pamelasoulis) [reply angular/src/app/modules/account/address/address.component.html]: Addressed these comments in #42

File: angular/src/app/modules/account/editable/editable.component.html [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+  
+  <input *ngIf="editMode" class="input level-item is-rounded swing-in-left-bck " name="userInput" ngModel type='text' [value]='data'
+    title="Edit and press Enter"
+    (keydown.enter)="onFocusOut($event)" 
+    (focusout)="onFocusOut($event)" 
+    [pattern]="pattern" #userInput="ngModel" placeholder={{data}}>
+  
+  
+  <span *ngIf="error" class="icon has-text-danger level-item">
+    <i class="fas fa-ban" title={{ErrorMessage}} aria-hidden=true></i>
```
[the remark this answers is not in the document]
Comment (pamelasoulis) [reply angular/src/app/modules/account/editable/editable.component.html]: Addressed in #42

File: angular/src/app/modules/account/editable/editable.component.html [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+  </span>
+  
+  <span *ngIf="valid" class="icon has-text-success level-item">
+    <i class="fas fa-check-square" aria-hidden=true title="Field was successfully edited"></i>
+  </span>
+  
+</div>
+
```
[the remark this answers is not in the document]
Comment (pamelasoulis) [reply angular/src/app/modules/account/editable/editable.component.html]: Also addressed in #42

File: angular/src/app/modules/account/editable/editable.component.ts [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * Class representing an _Editable Component_
+ */
+export class EditableComponent implements OnInit {
+  constructor() {}
+
+  @Input() data: string;
+  @Input() pattern: RegExp;
+  @Input() ErrorMessage: string;
+  @Input() Type: string;
```
[the remark this answers is not in the document]
Comment (pamelasoulis) [reply angular/src/app/modules/account/editable/editable.component.ts]: Fixed this in #42

File: angular/src/app/modules/account/payment/payment.component.ts (hunk 2 above)
Comment (pamelasoulis) [reply angular/src/app/modules/account/payment/payment.component.ts]: Fixed in #42

File: angular/src/app/modules/account/profile/profile.component.html (hunk 1 above)
Comment (pamelasoulis) [reply angular/src/app/modules/account/profile/profile.component.html]: Addressed this comment in #42

File: angular/src/app/modules/account/editable/editable.component.ts [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    if (this.Type) {
+      this.SetValidationType(this.Type);
+    }
+  }
+
+  /**
+   * Setter function for Validation Type
+   * @param Type string
+   */
+  public SetValidationType(Type: string): void {
```
[the remark this answers is not in the document]
Comment (pamelasoulis) [reply angular/src/app/modules/account/editable/editable.component.ts]: Addressed in #42

File: angular/src/app/modules/account/services/account.typeguard.ts (hunk 7 above)
Comment (qbob1) [reply angular/src/app/modules/account/services/account.typeguard.ts:3]: There was, I submitted #43 this weekend but it got buried from Freds latest. This is simply a means of avoiding dependencies

File: angular/src/app/modules/account/services/account.typeguard.ts (hunk 8 above)
Comment (qbob1) [reply angular/src/app/modules/account/services/account.typeguard.ts:18]: -.-

File: angular/src/app.config.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (TrueFlame42) [reply angular/src/app.config.json]: fixed

File: angular/src/app/modules/account/account-routing.module.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (TrueFlame42) [reply angular/src/app/modules/account/account-routing.module.ts:14]: fixed

File: angular/src/app/modules/account/editable/editable.validation.config.ts [hunk 17]
```diff
@@ -0,0 +1,37 @@
+interface PatternMessage {
+  pattern: RegExp;
+  ErrorMessage: string;
+}
+const ValidationConfig: { [key: string]: PatternMessage } = {};
```
[the remark this answers is not in the document]
Comment (pamelasoulis) [reply angular/src/app/modules/account/editable/editable.validation.config.ts]: Addressed camelCase and code smell in commit 06201c7

File: angular/src/app/modules/account/editable/editable.component.ts [hunk 18]
```diff
@@ -1,43 +1,68 @@
...
+   * street
+   * postal
+   * country
+   */
+  @Input() type: string;
   @Output() dataChange: EventEmitter<string> = new EventEmitter<string>();
-  editMode = false;
+
+  @Input() editMode: boolean;
+  @Output() editModeChange: EventEmitter<boolean> = new EventEmitter<boolean>();
```
Comment (escalonn) [inline angular/src/app/modules/account/editable/editable.component.ts:42]: fyi type annotations arent necessary here, it's inferred just like for the properties just below it; but you can leave them in

File: angular/src/app/modules/account/payment/payment.component.spec.ts [hunk 19]
```diff
@@ -1,7 +1,7 @@
 import { async, ComponentFixture, TestBed } from '@angular/core/testing';
 import { PaymentComponent } from './payment.component';
 import { AccountEditingService } from '../services/account.editing.service';
-
+import { Subject } from 'rxjs';
```
Comment (escalonn) [inline angular/src/app/modules/account/payment/payment.component.spec.ts:4]: newline after imports

File: angular/src/app/modules/account/payment/payment.component.spec.ts [hunk 20]
```diff
@@ -13,11 +13,14 @@ describe('PaymentComponent', () => {
   ];
   let component: PaymentComponent;
   let fixture: ComponentFixture<PaymentComponent>;
-
+  const AccountEditingServiceStub = {
```
Comment (escalonn) [inline angular/src/app/modules/account/payment/payment.component.spec.ts:16]: camelCase

File: angular/src/app/modules/account/services/abstract.editing.service.ts [hunk 21]
```diff
@@ -2,25 +2,38 @@ import { Subject } from 'rxjs';
...
   public editUpdates = this.subject.pipe(
     scan((acc, curr) => (typeof curr === 'object' ? Object.assign({}, acc, curr) : null), {})
   );
 
+  /**
+   * Piped from above, If this.typeguard() of above object is true, release it, else null
+   * skip first inorder to allow first update function to pass in correct model of type t
+   */
+
   public PayloadEmitter = this.editUpdates
```
Comment (escalonn) [inline angular/src/app/modules/account/services/abstract.editing.service.ts:22]: is there a reason for this to be TitleCase? same with Subject.
also, `public` is default, so why not be consistent with rest of code and leave it out.

File: angular/src/app/modules/account/services/account.editing.service.ts (hunk 5 above)
Comment (escalonn) [reply angular/src/app/modules/account/services/account.editing.service.ts:7]: previous review comment not addressed ^

File: angular/src/app/modules/account/services/account.type.ts [hunk 22]
```diff
@@ -0,0 +1,13 @@
...
+import { Address } from '../../../data/address.model';
+
+type AddressUpdate = {
+  address: Address;
+};
+type ProfileUpdate = {
+  profiles: Profile[];
+};
+
+export type IAccount = Account | ProfileUpdate | AddressUpdate;
```
Comment (escalonn) [inline angular/src/app/modules/account/services/account.type.ts:13]: prefer `interface` to `type`, and document what an "IAccount" is supposed to be. probably pick a better name too

File: angular/src/app/modules/account/account-routing.module.ts [hunk 23]
```diff
@@ -23,7 +24,12 @@ const routes: Routes = [{ component: AccountComponent, path: '' }];
     AccountReviewComponent,
     EditableComponent,
   ],
-  providers: [AccountEditingService],
+  providers: [
+    {
+      provide: GenericEditingService,
```
Comment (escalonn) [inline angular/src/app/modules/account/account-routing.module.ts:29]: pretty sure you should use an injectiontoken here instead, [like so](https://angular.io/guide/dependency-injection-providers#value-providers), to avoid the theoretical surprise when someone wants a different editingservice in this module but gets the account one. also add a brief comment either here or maybe in the genericeditingservice explaining that angular can't otherwise do generic service injection

File: angular/src/app/modules/account/editable/editable.component.html [hunk 24]
```diff
@@ -2,12 +2,12 @@
 
   <ng-content *ngIf="!editMode" select="*"></ng-content>
 
-  <input *ngIf="editMode" class="input level-item is-rounded swing-in-left-bck " name="userInput" ngModel type="text"
+  <input *ngIf="editMode" class="input level-item is-rounded swing-in-left-bck " name="userInput" type="text"
     [value]="data" [title]="titleInput" (keydown.enter)="onFocusOut($event)" (focusout)="onFocusOut($event)"
-    [pattern]="pattern" placeholder={{data}}>
+    [pattern]="pattern" placeholder="{{data}}">
```
Comment (escalonn) [inline angular/src/app/modules/account/editable/editable.component.html:7]: spaces inside `{{ }}` here and below

File: angular/src/app/modules/account/profile/profile.component.spec.ts [hunk 25]
```diff
@@ -27,7 +28,7 @@ describe('ProfileComponent', () => {
   beforeEach(async(() => {
     TestBed.configureTestingModule({
       declarations: [ProfileComponent],
-      providers: [{ provide: AccountEditingService, useValue: AccountEditingServiceStub }],
+      providers: [{ provide: GenericEditingService, useValue: AccountEditingServiceStub }],
```
Comment (escalonn) [inline angular/src/app/modules/account/profile/profile.component.spec.ts:31]: camelCase for the stub variable, and isn't its payloademitter property now wrong?

File: angular/src/app/services/editable/generic-editing.service.ts [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+import { Subject } from 'rxjs';
+import { scan, skip } from 'rxjs/operators';
+
+@Injectable()
+/**
+ * A lightweight state management class
+ * Accumulates objects/releases on set conditions via pipes
+ * Although its generic, T is best as a partial
+ * This allows sub objects of initial objects to overwrite
+ * state of the accumulator object
```
Comment (escalonn) [inline angular/src/app/services/editable/generic-editing.service.ts:11]: looks good. if it actually needs to be a partial then consider replacing `T` with `Partial<T>` in the class members, but if there's theoretical reason to not use partial then it's fine as is

File: angular/src/app/data/edited-account.type.ts [hunk 27]
```diff
@@ -0,0 +1,12 @@
+import { Account } from './account.model';
+
+type Partial<T> = {
+  [P in keyof T]?: T[P];
+};
```
[the remark this answers is not in the document]
Comment (qbob1) [reply angular/src/app/data/edited-account.type.ts:5]: That's what I thought... but my reference was: 
https://www.typescriptlang.org/docs/handbook/advanced-types.html
I guess the docs just reference the source

File: angular/src/app/services/editable/generic-editing.service.ts (hunk 26 above)
Comment (qbob1) [reply angular/src/app/services/editable/generic-editing.service.ts:11]: In theory if someone wanted a strict overwrite object then partial wouldn't be acceptable.

File: angular/src/app/modules/account/account-routing.module.ts [hunk 28]
```diff
@@ -26,8 +25,13 @@ const routes: Routes = [{ component: AccountComponent, path: '' }];
...
   providers: [
     {
-      provide: GenericEditingService,
-      useFactory: () => new GenericEditingService<EditedAccount>(),
+      /**
+       * Injection token used:
+       * To allow for multiple editing services
+       * Comply with angular's injection
+       */
+      provide: 'EditingService',
```
Comment (escalonn) [inline angular/src/app/modules/account/account-routing.module.ts:33]: this is just a string token not an InjectionToken. this will work but involves magic string throughout several files without any build-time checking of e.g. typos, and also, "EditingService" is too general a name, another module might want to use that token too. so define a constant InjectionToken; here in this module file is fine, and refer to that token everywhere instead of 'EditingService' string

File: angular/src/app/services/editable/generic-editing.service.ts [hunk 29]
```diff
@@ -24,7 +24,13 @@ export class GenericEditingService<T> {
    * Filter payload conditions below
    */
   editUpdates = this.subject.pipe(
-    scan((acc, curr) => (typeof curr === 'object' ? Object.assign({} as T, acc, curr) : null), {})
+    scan(
+      (acc, curr) => (typeof curr === 'object' ? Object.assign({} as T, acc, curr) : null),
+      /**
+       * This Seed value could be updated to reflect an inital state
```
Comment (escalonn) [inline angular/src/app/services/editable/generic-editing.service.ts:30]: spelling.
also, there's no point in /** \n * \n */ type comments when they're on method contents - just use // if it fits on one line.

File: angular/src/app/modules/account/account-routing.module.ts (hunk 28 above)
Comment (escalonn) [reply angular/src/app/modules/account/account-routing.module.ts:33]: [docs](https://angular.io/guide/dependency-injection-in-action#injectiontoken-objects); properly it would look something like `new InjectionToken<GenericEditingService<Partial<Account>>>('AccountEditingService');`

Comment (qbob1) [reply angular/src/app/modules/account/account-routing.module.ts]: Token created. I needed to move it to a separate folder in order to avoid circular deps. Please advise on structure of current file.

File: angular/src/app/modules/account/account-editing.token.ts [hunk 30]
```diff
@@ -0,0 +1,8 @@
+import { GenericEditingService } from '../../services/editable/generic-editing.service';
+import { InjectionToken } from '@angular/core';
+export const ACCOUNT_EDITING_SERVICE = new InjectionToken<GenericEditingService<Partial<Account>>>(
+  'AccountEditingService',
+  {
+    factory: () => new GenericEditingService<Partial<Account>>(),
+  }
+);
```
Comment (qbob1) [inline angular/src/app/modules/account/account-editing.token.ts:8]: Please advise on the placement of this file, is currently in the account module folder

Comment (qbob1) [reply angular/src/app/modules/account/account-editing.token.ts:8]: This looks like it may need to be provided in the module..

Comment (escalonn) [reply angular/src/app/modules/account/account-editing.token.ts:8]: this is fine but all the tests should just use this token, rather than defining a different one, which i'd expect to not work. and yes this does need to be provided in either accountmodule or accountroutingmodule

File: angular/src/app/services/editable/generic-editing.service.ts [hunk 31]
```diff
@@ -26,9 +26,7 @@ export class GenericEditingService<T> {
   editUpdates = this.subject.pipe(
     scan(
       (acc, curr) => (typeof curr === 'object' ? Object.assign({} as T, acc, curr) : null),
-      /**
-       * This Seed value could be updated to reflect an inital state
-       */
+      // This Seed value could be updated to reflect an inital state
```
Comment (escalonn) [inline angular/src/app/services/editable/generic-editing.service.ts:29]: i already said spelling in last review

File: angular/src/app/modules/account/account-routing.module.ts [hunk 32]
```diff
@@ -22,7 +24,12 @@ const routes: Routes = [{ component: AccountComponent, path: '' }];
     AccountReviewComponent,
     EditableComponent,
   ],
-  providers: [],
+  providers: [
+    {
+      provide: ACCOUNT_EDITING_SERVICE,
+      useFactory: () => new GenericEditingService<Partial<Account>>(),
+    },
```
Comment (escalonn) [inline angular/src/app/modules/account/account-routing.module.ts:31]: no need to specify the factory in both places, afaik: pick one or the other; if you pick the other then make this just `providers: [ACCOUNT_EDITING_SERVICE]`

File: angular/src/app/modules/account/profile/profile.component.spec.ts [hunk 33]
```diff
@@ -1,9 +1,7 @@
 import { async, ComponentFixture, TestBed } from '@angular/core/testing';
 import { Subject } from 'rxjs';
 import { ProfileComponent } from './profile.component';
-import { GenericEditingService } from '../../../services/editable/generic-editing.service';
-import { NgModule, InjectionToken } from '@angular/core';
-
+import { ACCOUNT_EDITING_SERVICE } from '../account-editing.token';
```
Comment (escalonn) [inline angular/src/app/modules/account/profile/profile.component.spec.ts:4]: newline after imports; unused import Subject i think

File: angular/src/app/modules/account/account-routing.module.ts (hunk 32 above)
Comment (qbob1) [reply angular/src/app/modules/account/account-routing.module.ts:31]: needed to use useExisting
